WalletWasabi.Tests/UnitTests/EventSourcing/InMemoryEventRepositoryTests.cs

```diff
@@ -0,0 +1,452 @@
+	{
+		private const string TestRoundAggregate = "TestRoundAggregate";
+		private readonly TimeSpan _semaphoreWaitTimeout = TimeSpan.FromSeconds(1);
+
+		public InMemoryEventRepositoryTests(ITestOutputHelper output)
+		{
+			Output = output;
+			TestEventRepository = new TestInMemoryEventRepository(output);
+			EventRepository = TestEventRepository;
+		}
+
+		private IEventRepository EventRepository { get; init; }
+		private TestInMemoryEventRepository TestEventRepository { get; init; }
+
+		private ITestOutputHelper Output { get; init; }
+
+		[Fact]
+		public async Task AppendEvents_Zero_Async()
+		{
+			// Arrange
+			var events = Array.Empty<WrappedEvent>();
+
+			// Act
+			await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events);
+
+			// Assert
+			Assert.True((await EventRepository.ListEventsAsync(nameof(TestRoundAggregate), "1"))
+				.SequenceEqual(events));
+			Assert.True((await EventRepository.ListAggregateIdsAsync(nameof(TestRoundAggregate)))
+				.SequenceEqual(Array.Empty<string>()));
+		}
+
+		[Fact]
+		public async Task AppendEvents_One_Async()
+		{
+			// Arrange
+			var events = new[]
+			{
+				new WrappedEvent(1),
+			};
+
+			// Act
+			await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events);
+
+			// Assert
+			Assert.True((await EventRepository.ListEventsAsync(nameof(TestRoundAggregate), "1"))
+				.SequenceEqual(events));
+			Assert.True((await EventRepository.ListAggregateIdsAsync(nameof(TestRoundAggregate)))
+				.SequenceEqual(new[] { "1" }));
+		}
+
+		[Fact]
+		public async Task AppendEvents_Two_Async()
+		{
+			// Arrange
+			var events = new[]
+			{
+				new WrappedEvent(1),
+				new WrappedEvent(2),
+			};
+
+			// Act
+			await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events);
+
+			// Assert
+			Assert.True((await EventRepository.ListEventsAsync(nameof(TestRoundAggregate), "1"))
+				.SequenceEqual(events));
+			Assert.True((await EventRepository.ListAggregateIdsAsync(nameof(TestRoundAggregate)))
+				.SequenceEqual(new[] { "1" }));
+		}
+
+		[Fact]
+		public async Task AppendEvents_NegativeSequenceId_Async()
+		{
+			// Arrange
+			var events = new[]
+			{
+				new WrappedEvent(-1)
+			};
+
+			// Act
+			async Task Action()
+			{
+				await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events!);
+			}
+
+			// Assert
+			(await Assert.ThrowsAsync<ArgumentException>(Action))
+				.Message.Contains("first event sequenceId is not natural number");
+		}
+
+		[Fact]
+		public async Task AppendEvents_SkippedSequenceId_Async()
+		{
+			// Arrange
+			var events = new[]
+			{
+				new WrappedEvent(2)
+			};
+
+			// Act
+			async Task Action()
+			{
+				await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events!);
+			}
+
+			// Assert
+			(await Assert.ThrowsAsync<ArgumentException>(Action))
+				.Message.Contains("invalid firstSequenceId (gap in sequence ids) expected: '1' given: '2'");
```

Comment:
This looks like a bug to me.

I would just use:

```suggestion
			ArgumentException argumentException = await Assert.ThrowsAsync<ArgumentException>(Action);
			Assert.Contains("invalid firstSequenceId (gap in sequence ids) expected: '1' given: '2'", argumentException.Message, StringComparison.Ordinal);
```

to actually test the "contains part".

Reply:
correct, I was using Moq extension and then it was bannished so I guess that is how it came to be